File: EditDistance/EditDistance.cpp
// EditDistance.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <string>

int GetMinOperations(std::string& s1, std::string& s2, size_t index1, size_t index2)
{
    if (index1 == s1.size())
    {
        return static_cast<int>(s2.size() - index2);
    }
    if (index2 == s2.size())
    {
        return static_cast<int>(s1.size() - index1);
    }

    int minOperations = INT_MAX;
    if (s1[index1] == s2[index2])
        return GetMinOperations(s1, s2, index1 + 1, index2 + 1);

    int delChar = GetMinOperations(s1, s2, index1 + 1, index2);
    int insChar = GetMinOperations(s1, s2, index1, index2 + 1);
    int repChar = GetMinOperations(s1, s2, index1 + 1, index2 + 1);
    minOperations = std::min(delChar, minOperations);
    minOperations = std::min(insChar, minOperations);
    minOperations = std::min(repChar, minOperations);

    return minOperations + 1;
}

int MinOperations(std::string& s1, std::string& s2)
{
    return GetMinOperations(s1, s2, 0, 0);
}

int main()
{
    while (true)
    {
        std::cout << "String 1: ";
        std::string s1;
        std::cin >> s1;
        if (s1 == "q" || s1 == "Q")
            break;

        std::cout << "String 2: ";
        std::string s2;
        std::cin >> s2;

        std::cout << "The levenshtein distance is: " << MinOperations(s1, s2) << std::endl;
    }
}

File: Go/EditDistance.go
package main

import "fmt"

func GetMinOperations(s1, s2 string, i, j int) int {
	if i == len(s1) {
		return len(s2) - j
	}
	if j == len(s2) {
		return len(s1) - i
	}

	if s1[i] == s2[j] {
		return GetMinOperations(s1, s2, i+1, j+1)
	}

	minOp := 32767
	minOp = min(minOp, GetMinOperations(s1, s2, i+1, j))
	minOp = min(minOp, GetMinOperations(s1, s2, i, j+1))
	minOp = min(minOp, GetMinOperations(s1, s2, i+1, j+1))
	return minOp + 1
}

func minOperations(s1, s2 string) int {
	return GetMinOperations(s1, s2, 0, 0)
}

func main() {
	for {
		fmt.Print("Enter string 1: ")
		var s1 string
		fmt.Scanln(&s1)
		if s1 == "q" || s1 == "Q" {
			break
		}

		fmt.Print("Enter string 2: ")
		var s2 string
		fmt.Scanln(&s2)

		n := minOperations(s1, s2)
		fmt.Printf("The levenshtein distance between %s and %s is %d\n", s1, s2, n)
	}
}
